fossilize_db.cpp

```diff
@@ -27,6 +27,15 @@
 #include <fcntl.h>
 #endif
 
+#ifdef __linux__
```

Comment:
Overall the fadvise calls seem reasonable, although I have some concerns. This is a pretty obscure corner of Linux, so:

- How does DONTNEED affect other processes? Do all processes using a particular range of the file have to call DONTNEED or close the file for it to be effective?

- The ranges here seem very narrow. Like a few bytes each time, and each blob tends to be rather small. Is it even meaningful to use fadvise on tiny byte ranges like this? Can Linux somehow link together these DONTNEED requests into complete blocks/pages?

- Looking at the man-page there is POSIX_FADV_NOREUSE. Isn't this more appropriate in some scenarios?

- Is there any way to observe the page cache behavior anywhere so we're not just adding this blindly? I think is the key, since I can play around with this myself while I'm thinking how to refactor this part of Fossilize into something more efficient. To make this work well, I feel there will be a fair amount of tuning needed.

The access pattern is going to be something like where the primary process first scans through the archive.

This is fully sequential. The entire archive isn't actually read, **only the metadata**, but a sequential scan should prime everything, and that seems to be the behavior we're getting by default on Linux. This step serves two purposes, figuring out how to subdivide work for the children, and warm the page cache for child processes later. 
https://github.com/ValveSoftware/Fossilize/blob/master/cli/fossilize_replay_linux.hpp#L455

Child processes come in and also scan through the archive in a strided fashion to figure out the metadata. This is technically redundant, and I have some ideas how to fix this. We cannot rely on fork() on Windows, so most likely I will emit a prepared metadata block in SHM instead, so that child processes do not have to parse anything. If the disk cache is primed, the current implementation **should** be fast, although I can see a problem if the archive is just too large (many many GB), it's going to thrash in some way.

After the initial scan, and when replay begins, the access pattern turns quite random as blobs are read as needed and out of order. Not all blobs are read, especially pipeline objects are subdivided among child processes to process. This still works well if the entire archive is present in the page cache. I think the current code only emits DONTNEED hints when blobs are actually read, but we should probably ensure that unused parts of the file don't get cached? I'm not sure how SEQUENTIAL works here. If it implicitly de-priorities data once it's read, that's probably fine.

As for writes, Fossilize writes very little data on its own. The only writes I would worry about is the disk cache writers in the drivers.

Reply:
> Overall the fadvise calls seem reasonable, although I have some concerns. This is a pretty obscure corner of Linux, so:

Yep.

> * How does DONTNEED affect other processes? Do all processes using a particular range of the file have to call DONTNEED or close the file for it to be effective?

`DONTNEED` operates on ranges in the cache. If there's dirty data in the range, the request is ignored, otherwise the cache is discarded - no matter which process was using it in the first place. But I don't think that's a concern here because fossilize is the only user of the data at that time.

> * The ranges here seem very narrow. Like a few bytes each time, and each blob tends to be rather small. Is it even meaningful to use fadvise on tiny byte ranges like this? Can Linux somehow link together these DONTNEED requests into complete blocks/pages?

This is a valid concern. Maybe we'd need to ask kernel guys or look at the kernel source. I could imaging that is discards all pages touched by the range - unless those pages a dirty where it would fall back to ignoring the request.

> * Looking at the man-page there is POSIX_FADV_NOREUSE. Isn't this more appropriate in some scenarios?

In theory, yes. But (a) it seems to be made for the write case, and (b) it isn't implemented at all in current kernels. The man page actually states that, and it says previous implementation semantics have been a bug. There's no cache hinting for writes in Linux.

> * Is there any way to observe the page cache behavior anywhere so we're not just adding this blindly? I think is the key, since I can play around with this myself while I'm thinking how to refactor this part of Fossilize into something more efficient. To make this work well, I feel there will be a fair amount of tuning needed.

Given that you keep the rest of your system idle during a test, you can drop caches to prepare it as a cold run and then watch `egrep '(Cached|Dirty|Write):' /proc/meminfo` before and after running fossilize-replay. It will give a rough idea of how effectively we freed cache. You could also run fossilize-replay in a cgroup with memory accounting turned on and watch how much memory it allocated: This accounting will also include cache usage. The latter is what I did (tho, I looked at the Steam cgroup so I didn't test fossilize isolated), and looking at how fast swap usage increased. With these patches, my system stood much more responsive during the 21 GB fossilize job when running the Path of Exile shaders. Apparently, responsiveness is hard to measure, it's subjective.

> The access pattern is going to be something like where the primary process first scans through the archive.
> 
> This is fully sequential. The entire archive isn't actually read, **only the metadata**, but a sequential scan should prime everything, and that seems to be the behavior we're getting by default on Linux. This step serves two purposes, figuring out how to subdivide work for the children, and warm the page cache for child processes later.
> https://github.com/ValveSoftware/Fossilize/blob/master/cli/fossilize_replay_linux.hpp#L455

Yeah but actually we probably don't want to prime the cache - this makes 9 GB for the Path of Exile shaders worth of cache - which is probably discarded early if it is that huge. This doesn't even fit within the Average Joe system with 8-12 GB of RAM. Instead, we could use fadvise to tell the kernel that we are doing sequential access (which will double the read-ahead window for this file), or that we are doing random access (which will set read-ahead to 0). And if we know we are not going to need that data again soonish, we should discard it because we are not a foreground process.

> Child processes come in and also scan through the archive in a strided fashion to figure out the metadata. This is technically redundant, and I have some ideas how to fix this.

Yeah, I think that's a good slice of the problem.

> We cannot rely on fork() on Windows,

Apparently. Makes me sad. :-(

> so most likely I will emit a prepared metadata block in SHM instead, so that child processes do not have to parse anything. If the disk cache is primed, the current implementation **should** be fast, although I can see a problem if the archive is just too large (many many GB), it's going to thrash in some way.

The biggest problem is when the processes have to copy that data into private memory - no matter if it's shared or isn't. If we exhaust the system memory by that way, there's not even a point in trying to be as disk friendly as possible because we are going to flush the cache from memory. In that case, we don't need to prime anything in the first place. Also, the children are not latency sensitive, I don't think priming serves any valuable purpose.

> After the initial scan, and when replay begins, the access pattern turns quite random as blobs are read as needed and out of order. Not all blobs are read, especially pipeline objects are subdivided among child processes to process.

In that case we should use `FADV_RANDOM` to turn off read-ahead: We don't need to read additional 64 kB if we only going to read some small randomly placed blobs. If IO priority is lowered, the system can take care of reordering IO requests optimally so we don't need to prime the cache to reduce seek overhead. At least for Linux. I think Windows works very differently here.

> This still works well if the entire archive is present in the page cache. I think the current code only emits DONTNEED hints when blobs are actually read, but we should probably ensure that unused parts of the file don't get cached? I'm not sure how SEQUENTIAL works here. If it implicitly de-priorities data once it's read, that's probably fine.

As pointed out above: SEQUENTIAL doubles the read-ahead window (the kernel will read ahead 128k instead of 64k), essentially speeding up sequential workloads by reducing IO requests count. RANDOM sets read-ahead to 0, telling the kernel that we exactly only want to read that particular data and no read-ahead for adjacent data which we probably don't need at all.

In `strace`, it looks like the file stream function break big reads into small 4k pieces, so SEQUENTIAL made sense here.

> As for writes, Fossilize writes very little data on its own. The only writes I would worry about is the disk cache writers in the drivers.

Yeah, it was hard for me to spot any write at all during `strace`. All the writes seem to come from the nvidia driver lib. But it opens two file descriptors on behalf the process as can be seen in `/proc/<fossilize_pid>/fd`. It should actually be possible to issue `fsync()` or `fadvise` on those descriptors. But for a first implementation, it should be enough to use `fsync()` every now and then (maybe by looking at iostats and sync on every 4 MB written): It should block writes on the fd until the write-back is drained and thus throttle the write-back to a sane value. Background info: Linux tends to stall on writes because it considers blocking reads more perceivable to interactive users than blocking writes: After all, you'd expect the application to stall when you save your document. So if we put a file descriptor into synchronize state, we should actually successfully throttle accumulation of dirty data in the cache, preventing the kernel from swapping too early.

So, yeah, I also concluded that we only need to worry about the writes to files we actually cannot control. And accumulation of dirty data in the cache is one of the remaining problems I'm seeing (except the duplication of memory anyways), and it will write back only slowly because we are reading a lot of stuff and Linux prefers reading over writing. At some point, it will start blocking all processes doing IO just get rid of the pile of dirty writeback data - and then your systems stalls very visibly and may freeze for seconds or even minutes.

Reply:
BTW: I was thinking to not only consider HW threads for how many children to spawn, maybe also consider `ceil(sysmem * 0.5 / dbsize)` and use the lowest of both values, or some biased value like `(2 * sysmem_count + hwthreads_count) / 3`... But the memory duplication should be fixed in the first place. And also think about datasets which do not fit into sysmem at all. The Path of Exile dataset seems to be a good example, it's 9 GB for me.

Reply:
> > * The ranges here seem very narrow. Like a few bytes each time, and each blob tends to be rather small. Is it even meaningful to use fadvise on tiny byte ranges like this? Can Linux somehow link together these DONTNEED requests into complete blocks/pages?
> 
> This is a valid concern. Maybe we'd need to ask kernel guys or look at the kernel source. I could imaging that is discards all pages touched by the range - unless those pages are dirty where it would fall back to ignoring the request.

I've asked in IRC and while I got no definitive answer, I was told this: Unless the request returns EINVAL, it'll also discard partially requested pages from the cache. So, if we check the return value, we'll know.

Reply:
@HansKristian-Work I've looked at the kernel. This is how it works: WILLNEED will round the range up to the outer page boundary, DONTNEED will round the range to the inner page boundary. So we cannot DONTNEED sub-page ranges.

It's here:
https://github.com/torvalds/linux/blob/96485e4462604744d66bf4301557d996d80b85eb/mm/fadvise.c#L68

Reply:
According to kernel source, DONTNEED on dirty cache will trigger immediate writeback flushing. This is different to what the man pages states where it says "the implementation MAY do it". I think that's POSIX stuff... Linux does it, POSIX doesn't require it.

Reply:
> If we exhaust the system memory by that way, there's not even a point in trying to be as disk friendly as possible because we are going to flush the cache from memory.

The metadata block is not going to be huge, a few bytes per entry. If *that* block is going to blow out anything, then you have other problems.

Reply:
> /proc/<fossilize_pid>/fd. It should actually be possible to issue fsync() or fadvise on those descriptors

We cannot rely on random driver implementations like this, Mesa works very differently, the only thing we can do is a more global fsync().

Reply:
>  At some point, it will start blocking all processes doing IO just to get rid of the pile of dirty writeback data.

I think that's the best argument for doing some kind of fsync() and IO throttling, but I have no idea how to even do this programmatically.

Reply:
Bear with me, I'm still in the process of understanding how it works. But the picture slowly becomes clear. Thanks for your patience.

Yeah, I don't think the individual metadata blocks do a lot of pressure. It's the sum of all small pieces for a big file. DONTNEED at close time seems to help. And yes, the own copy currently seems the biggest problem but I cannot help here a lot, you know the code much better.

I'd probably go for the write pressure introduced by the drivers own shader cache next. Could you point me to where the writes by the driver are probably injected? Should be somewhere around the compiler calls, right?

Reply:
vkCreateGraphicsPipeline and vkCreateComputePipeline would be the place the driver does something, but it's not going to be in the calling thread. All drivers have a "disk cache thread" or similar, so it's completely asynchronous when this happens. I think what you can look into in that case is to check the IO writeback cache size and flush it.

The ideal place for this would be a thread in the main process, assuming it's possible to control fsync behavior for the entire process tree like this. If not, then somewhere here would be a good fit: https://github.com/ValveSoftware/Fossilize/blob/master/cli/fossilize_replay.cpp#L2086 We just kicked off some parsing work, and while that happens, we can poke at the IO cache a bit perhaps.

Reply:
> And yes, the own copy currently seems the biggest problem but I cannot help here a lot, you know the code much better.

Yes, I'll look into this soon. Should be a pretty good win in general.

Reply:
When looking at htop while the replayer starts writing, you can press L on the process to see the open files: The process has open file handles on the NVIDIA shader cache which I could access through `/proc/self/fd` and use `fadvise()` or `fsync()` on it. Not sure for mesa, tho. While it is probably async, the handles are visible in our process.

Reply:
Mesa uses one individual file per blob entry, so we won't be able to hook any individual FDs.